volume_type_validation/volume_type_validation.py

```diff
@@ -0,0 +1,240 @@
+                                                              group_volumes,
+                                                              group_volume_attachments,
+                                                              container_nodes_group)
+
+                recommendation['recommended_volume_type'] = \
+                    self._volume_type_id_to_str(recommended_volume_type)
+                recommendation['wrong_volume_type_vms'] = bad_hosts
+                recommendation['wrong_volume_type_volumes'] = bad_volumes
+                message['recommendations'].append(recommendation)
+
+            self.result.add_recommendations(source_id=str(source_id),
+                                            message=message)
+
+    def _find_bad_hosts(
+            self,
+            recommended_volume_type,
+            group_volumes,
+            group_volume_attachments,
+            container_nodes_group
+    ):
+        # Indentify volumes and their hosts that should change the volume type
+        bad_volumes = \
+            group_volumes[group_volumes['volume_type_id'] != recommended_volume_type]
+        bad_volume_attachments = \
+            group_volume_attachments[group_volume_attachments['volume_id'].isin(bad_volumes['id'])]
+        bad_nodes = \
+            container_nodes_group[container_nodes_group['lives_on_id'].isin(bad_volume_attachments['vm_id'])]
+        bad_hosts = self.vms[self.vms['id'].isin(bad_nodes['lives_on_id'])]
+
+        bad_hosts = bad_hosts[['id', 'name', 'source_ref']].to_dict('records')
+        bad_volumes = self._enrich_with_volume_type(
+            bad_volumes[['id', 'name', 'source_ref', 'volume_type_id']].to_dict('records'))
+
+        bad_volumes_array = pluck(bad_volumes, 'id')
+
+        bad_vm_to_bad_volume_connection = \
+            group_volume_attachments.loc[group_volume_attachments['volume_id'].isin(bad_volumes_array)]
+
+        for bad_volume in bad_volumes:
+            bad_volume['vm_id'] = \
+                bad_vm_to_bad_volume_connection.loc[
+                    bad_vm_to_bad_volume_connection['volume_id'] ==
+                    bad_volume['id']].vm_id.values[0]
+
+
+        return bad_volumes, bad_hosts
+
+    def _enrich_with_volume_type(self, items):
+        for value in items:
+            value['volume_type'] = \
+                self._volume_type_id_to_str(value['volume_type_id'])
+        return items
+
+    def _recommendations(self, host_id):
+        recommendations = []
+
+        clusters = self.result.invalid_items.keys()
+
+        for cluster in clusters:
+            for recommendation in self.result.invalid_items[cluster][0]['recommendations']:
+                bad_hosts = recommendation['wrong_volume_type_vms']
+                bad_volumes = recommendation['wrong_volume_type_volumes']
+                bad_host_match = list(filter(lambda bad_host: bad_host['id'] == host_id, bad_hosts))
```

Comment:
cleaner way to do this:
```suggestion
                bad_host_match = [h for h in bad_hosts if h['id'] == host_id]
```

Reply:
done.
Addressed in 8fb9d5ad5fc223986f95be62e3f447f958f5d9f6